### DISCUSSION TITLE
Label text word wrap

### DISCUSSION THREAD

**Discussion by TouwaStar** [excaliburjs/Excalibur#2458] [Q&A] [answered]
Is there a way to make text or label objects automatically word-wrap when reaching a certain size?

I want to define a sort of a textbox. 
My current solution just uses a Label and an Actor serving as the backdrop but even if I put them in a parent-child relation it does not seem to create any word wrapping.

Fitting Text
![image](https://user-images.githubusercontent.com/30479449/184260557-aa3d8865-d02b-4a55-a7f3-5dfaa38db7bd.png)

Not fitting text
![image](https://user-images.githubusercontent.com/30479449/184260679-06a416ef-8e0f-452a-97ee-1564ccdd89a1.png)

What automatic word-wraping would do
![image](https://user-images.githubusercontent.com/30479449/184260843-b86cb1b8-0b47-42d4-8809-d9157b909339.png)

Does the framework offer any way to tackle this ?

**Comment by eonarheim** [comment, marked as answer]
@TouwaStar Thanks for the discussion thread! Currently there isn't anything built into Excalibur to handle text word wrap. 

It's definitely something that would be a useful feature in Excalibur though! It is a tricky problem to solve for sure, I'll write up and issue around this so it can be supported 👍 

A couple workarounds I can think of:

1. The lo-fi: Manually manage the breaks and insert newline characters into your text`\n` and manually break into multiple lines. 
2. The hi-fi: Use an `ex.Canvas` object to manually draw text and use a canvas text library, this one https://github.com/geongeorge/canvas-txt looks pretty good and seems to work, example below:

<img width="328" alt="image" src="https://user-images.githubusercontent.com/612071/184512096-b884bef5-cfe5-4ddb-9ce4-217ca0cbbbe0.png">


```typescript
// <script src="//unpkg.com/canvas-txt"></script>
var canvasTxt = window.canvasTxt.default;
// or npm install canvas-txt --save
// import canvasTxt from 'canvas-txt'


var textGraphic = new ex.Canvas({
  width: 200,
  height: 200,
  draw: (ctx: CanvasRenderingContext2D) => {
    const txt = 'Some extra long text that should wrap properly to the next line automatically';
    ctx.save();
    canvasTxt.font = 'Verdana';
    canvasTxt.fontSize = 20;
    canvasTxt.align = 'left';
    // canvasTxt.debug = true //shows debug info
    canvasTxt.justify = false
    canvasTxt.drawText(ctx, txt, 0, 0, 200, 200);
    ctx.restore();
  }
});
var textActor = new ex.Actor();
textActor.graphics.use(textGraphic);

var game = new ex.Engine({
  width: 400,
  height: 400
});
game.currentScene.camera.pos = ex.vec(0, 0);

game.add(textActor);

game.start();
```

**Reply by TouwaStar** [reply]
i ended up doing a small function that will format the text for me to fit the box its in 
```
function fitTextInDiplayBox(box: Actor, label: Label) {
    var wordsInLabelText = label.text.split(" ");
    label.text = "";

    while (wordsInLabelText.length > 0) {
        var word = wordsInLabelText.shift();
        if (word === undefined) {
            break;
        }

        if (wordsInLabelText.length > 0) {
            word += " ";
        }

        label.text += word;

        if (label.getTextWidth() > box.width) {
            label.text = label.text.slice(0, -word.length)
            label.text += "\n";
            label.text += word;
        }
    }
}
```

**Reply by eonarheim** [reply]
@TouwaStar Awesome! Do you mind if we reference your snippet as a starting point for an implementation?

I've created an issue for Excalibur to support this type of thing out of the box https://github.com/excaliburjs/Excalibur/issues/2463

**Reply by TouwaStar** [reply]
sure, go ahead !

**Comment by TouwaStar** [comment]
Go ahead!

**Comment by djcsdy** [comment]
You might also be interested in [excalibur-extended-label](https://github.com/dcgw/excalibur-extended-label) (for Excalibur 0.24.x) and [excalibur-text](https://github.com/dcgw/excalibur-text) (for Excalibur 0.25+).

These are built specifically for our needs and have some other differences to the built-in Excalibur text rendering, but they do also support word wrap.

Other differences:
 * No support for `SpriteFont`.
 * `Font` is not used, font family, bold, etc are set directly on the `Graphic`/`Label` (necessary due to the different approach we took to rendering text).
 * Outline is drawn differently (drawn on the outside of the text instead of on the edge and partially inside the text (with some caveats if `color.alpha !== 1`, although you probably want to set `opacity` instead).
 * Shadow is drawn differently to avoid the shadow appearing over the text in some cases, although it should look the same as Excalibur's shadow in most cases.

**Reply by eonarheim** [reply]
Thanks for sharing! I'll definitely be checking these out for ideas if you're cool with that

**Reply by djcsdy** [reply]
Absolutely. I released it under the ISC license which is pretty unrestrictive, but please consider this comment a personal license for everyone who has contributed to excalibur to modify, reuse and/or redistribute the code of excalibur-extended-label and excalibur-text under any license you like, with or without attribution.
